daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/Convert.hs

```diff
@@ -1208,28 +1208,35 @@ toCtors env t = Ctors (getName t) <$> mapM convTypeVar (tyConTyVars t) <*> cs
 --   this would considerably complicate converting coercions. Second, it makes
 --   unpacking them faster in the current intepreter. Since the constructor
 --   names of newtypes are meant to be meaningless, this is acceptable.
+--
+-- * We add a field for every constraint containt in the thetas.
 ctorLabels :: TyConFlavour -> DataCon -> [FieldName]
-ctorLabels flv con
-  | flv `elem` [ClassFlavour, TupleFlavour Boxed] || isTupleDataCon con
-    -- NOTE(MH): The line below is a workaround for ghc issue
-    -- https://github.com/ghc/ghc/blob/ae4f1033cfe131fca9416e2993bda081e1f8c152/compiler/types/TyCon.hs#L2030
-    || (is con == "Unit" && moduleNameString (GHC.moduleName (nameModule (getName con))) == "GHC.Tuple")
-  = map (mkField . (:) '_' . show) [1..dataConSourceArity con]
-  | flv == NewtypeFlavour && null lbls
-  = [mkField "unpack"]
-  | otherwise
-  = map convFieldName lbls
-  where lbls = dataConFieldLabels con
+ctorLabels flv con =
+    [mkField $ "f" <> show i | i <- [1 .. length thetas]] ++ conFields flv con
```

Comment:
This will cause trouble when there's already a field of name `f1`. I propose we name those fields `$dict1` and so on. Users can't put a `$` in the name themselves.

Reply:
renamed.